blockchain/dkg/dkgtasks/dispute_share_distribution_task.go

```diff
@@ -0,0 +1,212 @@
+
+	if t.State.Phase != objects.DisputeShareDistribution && t.State.Phase != objects.ShareDistribution {
+		return fmt.Errorf("%w because it's not DisputeShareDistribution phase", objects.ErrCanNotContinue)
+	}
+
+	var participantsList = t.State.GetSortedParticipants()
+	// Loop through all participants and check to see if shares are valid
+	for idx := 0; idx < t.State.NumberOfValidators; idx++ {
+		participant := participantsList[idx]
+
+		var emptyHash [32]byte
+		if participant.DistributedSharesHash == emptyHash {
+			continue
+		}
+
+		logger.Infof("participant idx: %v:%v:%v\n", idx, participant.Index, t.State.Index)
+		valid, present, err := math.VerifyDistributedShares(t.State, participant)
+		if err != nil {
+			// A major error occured; we cannot continue
+			logger.Errorf("VerifyDistributedShares broke; Participant Address: %v", participant.Address.Hex())
+			return fmt.Errorf("VerifyDistributedShares broke: %v; %v", err.Error(), objects.ErrCanNotContinue)
+		}
+		if !present {
+			logger.Warningf("No share from %v", participant.Address.Hex())
+			continue
+		}
+		if !valid {
+			logger.Warningf("Invalid share from %v", participant.Address.Hex())
+			t.State.BadShares[participant.Address] = participant
+		}
+	}
+
+	return nil
+}
+
+// DoWork is the first attempt at disputing distributed shares
+func (t *DisputeShareDistributionTask) DoWork(ctx context.Context, logger *logrus.Entry, eth interfaces.Ethereum) error {
+	return t.doTask(ctx, logger, eth)
+}
+
+// DoRetry is subsequent attempts at disputing distributed shares
+func (t *DisputeShareDistributionTask) DoRetry(ctx context.Context, logger *logrus.Entry, eth interfaces.Ethereum) error {
+	return t.doTask(ctx, logger, eth)
+}
+
+func (t *DisputeShareDistributionTask) doTask(ctx context.Context, logger *logrus.Entry, eth interfaces.Ethereum) error {
+	t.State.Lock()
+	defer t.State.Unlock()
+
+	logger.Info("DisputeShareDistributionTask doTask()")
+
+	callOpts := eth.GetCallOpts(ctx, t.State.Account)
+
+	for _, participant := range t.State.BadShares {
+
+		isValidator, err := eth.Contracts().ValidatorPool().IsValidator(callOpts, participant.Address)
+		if err != nil {
+			return dkg.LogReturnErrorf(logger, "getting isValidator failed: %v", err)
+		}
+
+		if !isValidator {
+			continue
+		}
+
+		dishonestAddress := participant.Address
+		encryptedShares := t.State.Participants[participant.Address].EncryptedShares
+		commitments := t.State.Participants[participant.Address].Commitments
+
+		// Construct shared key
+		disputePublicKeyG1, err := bn256.BigIntArrayToG1(participant.PublicKey)
+		if err != nil {
+			return err
+		}
+		sharedKeyG1 := cloudflare.GenerateSharedSecretG1(t.State.TransportPrivateKey, disputePublicKeyG1)
+		sharedKey, err := bn256.G1ToBigIntArray(sharedKeyG1)
+		if err != nil {
+			return err
+		}
+
+		// Construct shared key proof
+		g1Base := new(cloudflare.G1).ScalarBaseMult(common.Big1)
```

Comment:
Check if this section has a higher level function in crypto to use, consult chris if needed.